### ISSUE TITLE
Cache normalisation with nested type

### ISSUE THREAD

**Issue by yliu342** [apollographql/apollo-ios#2162]
Hi, I've followed the link about [client caching](https://www.apollographql.com/docs/ios/caching/#controlling-normalization) and have my cache implemented. I've set the cacheKeyForObject as 
```
cacheKeyForObject = { [$0["__typename"], $0["id"]] }
``` 
However I noticed when fetch data from Cache any sub type which does not have an `ID` field will have exactly the same value.

Take the following as example, I have 2 class `Book` and `Author`
```
public struct Book: {
  var id: String
  var name: String
  var author: [Author]
}

public struct Author: {
  var name: String
  var dob: String
}
```
When looking at the SQL cache for a query for books, I can see the query for Books is being cached with key `Optional(Book)_Optional({what_ever_id})`. With value `{id:xxx, name:xxx, author:[($reference: Optional(Author)_nil)]}`. 
The value for `Optional(Author)_nil)` will be the author JSON from the last book type. 
When I get value from cache, if there are multiple `Book` objects, all Book object will have the same Author array.  Since the `Author` object doesn't have an ID field is it possible to use the parent ID? In this case use the Book's ID?

**Comment by yliu342** [comment]
I think I figured it out. What I did wrong was my `cacheKeyForObject`. I had it set `cacheKeyForObject = { [$0["__typename"], $0["id"]] }` this works for all type that has an `id` field. But for type that don't have an `id` field it will use `__typename_nil`. 

Now I've changed my cacheKeyForObject to 
```
cacheKeyForObject = {
            if $0["id"] == nil {
                return nil
            }
            return [$0["__typename"], $0["id"]]
        }
```
If the object does not have `id` field simply return `nil` and it will use the path as they key and this solved my problem.
